### PR TITLE
`Exam mode`: improve attendance check with websocket based id information

### LINKED ISSUE #7454: `Exam mode`: Improve attendance check with websocket based Id information
### Is your feature request related to a problem?

The on site exam mode requires the check of the matriculation number, the name of the student and the artemis username. This information are used to savely identify the student and the open artemis session on the student device. 
Currently the student has to open a new tab and show his credentials. This should be improved and implemented more interactively. 

### Describe the solution you'd like

The instructor should be able to initiate a pop up from the instructors iPad that opens on the students enddevice in the ongoing exam mode. 
This pop up should contain the matriculation number, the name of the student and the artemis username. 
This information is needed to identify the student and his open exam session. 

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by max-bergmann**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [X] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [X] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [X] I documented the Java code using JavaDoc style.
#### Client
- [X] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [X] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [X] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [X] I documented the TypeScript code using JSDoc style.
- [X] I added multiple screenshots/screencasts of my UI changes.
- [X] I translated all newly inserted strings into English and German.

### Motivation and Context

Closes #7454

### Description

Closes #7454 

#### Exam Mode Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Exam

URL:
```
https://{serverurl}/api/courses/{courseId}/exams/{examId}/students/{studentLogin}/attendance-check
```
Example:
```
https://artemistest3.ase.in.tum.de/api/courses/206/exams/400/students/artemis_test_user_1/attendance-check
```

1. Log in to Artemis
2. Generate Exam as an instructor
3. Fill all required fields (1 Exercise is enough)
4. Participate in the exam as a student (2 sessions with 2 users)
5. Link Postman and fire specific Attendance Event - the api url is used from the iOS Attendance App
6. Make sure that the Event is shown
7. Make sure that the Event is not Listed in the exam notification list
8. Fire other events (Working-Time Update / Announcement)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Test Coverage

| Class/File                                | Line Coverage | Confirmation (assert/expect) |
|-------------------------------------------|--------------:|-----------------------------:|
| ExamLiveEventsService.java                |          100% | ✅                           |
| StudentExamResource.java                  |           85% | ✅                           |
| exam-participation-live-events.service.ts |        90,08% | ✅                           |
| exam-live-events-button.component.ts      |        96,29% | ✅                           |
| exam-live-events-overlay.component.ts     |        91,89% | ✅                           |
| exam-live-event.component.ts              |        92,85% | ✅                           |

### Screenshots

image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java [hunk 1]
```diff
@@ -234,6 +231,37 @@ public ResponseEntity<StudentExam> updateWorkingTime(@PathVariable Long courseId
...
+     * @return the ResponseEntity with status 200 (OK) and with the updated student exam as body
+     */
+    @PostMapping("/courses/{courseId}/exams/{examId}/students/{studentLogin:" + Constants.LOGIN_REGEX + "}/attendance-check")
+    @EnforceAtLeastInstructor
+    public ResponseEntity<ExamAttendanceCheckEventDTO> attendanceCheck(@PathVariable Long courseId, @PathVariable Long examId, @PathVariable String studentLogin,
+            @RequestBody(required = false) String message) {
+        log.debug("REST request for attendance-check for student : {}", studentLogin);
+
+        var student = userRepository.findOneWithGroupsAndAuthoritiesByLogin(studentLogin)
+                .orElseThrow(() -> new EntityNotFoundException("User with login: \"" + studentLogin + "\" does not exist"));
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:249]: We often have "ElseThrow" variants of the database retrievals. Can you also add one for this query and replace the other calls to this method as well?

File: src/main/webapp/i18n/de/exam.json [hunk 2]
```diff
@@ -182,6 +183,10 @@
                         "titlePersonal": "Deine persönliche Bearbeitungszeit dieser Klausur wurde geändert.",
                         "old": "Vorherige Bearbeitungszeit",
                         "new": "Neue Bearbeitungszeit"
+                    },
+                    "examAttendanceCheck": {
+                        "description": "Eine Klausur-Aufsicht hat die Anwesenheitskontrolle für Dich ausgelöst. Bitte folge den Anweisungen der Aufsicht.",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/exam.json:188]:
```suggestion
                        "description": "Eine Klausur-Aufsicht hat die Anwesenheitskontrolle für dich ausgelöst. Bitte folge den Anweisungen der Aufsicht.",
```

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by JohannesStoehr — APPROVED** [review]
Code looks good to me

**Review by MichaelOwenDyer — APPROVED** [review]
Code looks good

**Review by milljoniaer — DISMISSED** [review]
Tested on ts4, works as expected 👍

**Review by valentin-boehm — DISMISSED** [review]
Tested on TS4 👍

**Review by RY997 — APPROVED** [review]
Tested on TS4

File: src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java [hunk 3]
```diff
@@ -234,6 +231,39 @@ public ResponseEntity<StudentExam> updateWorkingTime(@PathVariable Long courseId
...
+    @EnforceAtLeastInstructor
+    public ResponseEntity<ExamAttendanceCheckEventDTO> attendanceCheck(@PathVariable Long courseId, @PathVariable Long examId, @PathVariable String studentLogin,
+            @RequestBody(required = false) String message) {
+        log.debug("REST request for attendance-check for student : {}", studentLogin);
+
+        var student = userRepository.getUserByLoginElseThrow(studentLogin);
+
+        StudentExam studentExam = studentExamRepository.findWithExercisesByUserIdAndExamId(student.getId(), examId).orElseThrow();
+
+        studentExam.getUser().setVisibleRegistrationNumber();
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/web/rest/StudentExamResource.java:252]: This is not needed, correct? You're only using the studentExam id further down

File: src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package de.tum.in.www1.artemis.domain.exam.event;
+
+import javax.persistence.Column;
+import javax.persistence.DiscriminatorValue;
+import javax.persistence.Entity;
+
+import de.tum.in.www1.artemis.web.rest.dto.examevent.ExamAttendanceCheckEventDTO;
+
+/**
+ * An event indicating an updated working time for a specific student exam.
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/exam/event/ExamAttendanceCheckEvent.java:10]: copied javadoc.

**Comment by Strohgelaender** [comment]
How does this event get triggerd? Only from the attendance check app? I don't see a way to fire this event as an instructor by looking at the code changes?

I think the long text "An exam supervisor has triggered the attendance check for you. Please follow the instructions of the supervisor." is unnecessary, especially considering the the responsible instructor should already by standing next to the student and be talking to the student when this popup shows up. I'd like to see that removed, but this can also be done in a followup considering the time pressure.

Your screenshot does not include the registration number, but it gets set when calling `own-student-exam`. Why is this the case? I think this information should be added to the popup.

File: src/main/webapp/app/exam/shared/events/exam-live-event.component.html [hunk 5]
```diff
@@ -23,6 +23,33 @@
...
+                    </tr>
+                    <tr>
+                        <th>{{ 'artemisApp.studentExamDetail.email' | artemisTranslate }}</th>
+                        <td>{{ event.user?.email }}</td>
+                    </tr>
+                    <tr *ngIf="event.user?.visibleRegistrationNumber">
+                        <th>{{ 'artemisApp.studentExamDetail.matriculationNumber' | artemisTranslate }}</th>
+                        <td>{{ event.user?.visibleRegistrationNumber }}</td>
+                    </tr>
+                    <tr *ngIf="examAttendanceCheckEvent.text && examAttendanceCheckEvent.text.length > 0">
```
Comment (Strohgelaender) [inline src/main/webapp/app/exam/shared/events/exam-live-event.component.html:46]: grafik

```suggestion
                    <tr *ngIf="examAttendanceCheckEvent.text">
```

**Review by Strohgelaender — APPROVED** [review]
code aproved, only one optional change:

**Pull request merged**